First draft of DOi design doc (#212)
dandi/dandi-api: doc/design/doi-generation-1.md

> # DOI generation process

Reviewer: ```suggestion
# Publish process

This process is predicated on a few tasks that are expected to receive valid input and assuming that only certain errors could happen that are out of our control.

- publish enabled only if dandiset is valid (so all validation checks have passed)
- on publish trigger:
   - create the versioned release in the db (this prevents any asset/blob belonging to the release from being deleted)
   - insert all the relevant metadata (version, doi, url, manifest location/checksum) into the metadata record of the new version 
  - send this new metadata record to mint the doi (there should be no reason why this should fail unless the external resource fails. if it fails notify admins since this is likely a new bug)
   - write out to s3 bucket (even on any datacite failure)
   - switch over ownership of the release to dandi (unless releases are not modifiable by the dandiset creator)
   
## DOI generation process
```

Author: I thought this PR would be scoped to DOIs so Dorota and I could get a cohesive design. If you want to cover the entire publish process, I can add:
* Metadata validation (so the publish button isn't available if metadata is invalid)
* Web GUI UX
* Creating the published version
* Writing the manifests

Reviewer: yes let's scope the whole thing. because the datacite conversion and doi generation sits within the context of that.

## dandi-api

The DOI generation process happens inside a call to `POST /dandisets/{dandiset_id}/draft/publish/`.
DOI generation is the first step of the publish process, because it is the only step to rely on an external service.
If the call to the Datacite API fails, there are no files in S3 that need to be cleaned up.

The API will pass the following to the DOI generator:

> * A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/000001/0.210402.1835`.

Reviewer 2: ```suggestion
* A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/dandi.000001.v0.210402.1835`.
```
Given the discussion elsewhere we might want to mention dandi and avoid /

* A URL for the dandiset. This will be a redirector URL of the form `f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'`.
* The dandiset metadata, as a python dictionary.

## dandi.core.doi ?

This function will generate the body of the DOI and return it.

> TODO: fill me in

Reviewer 3: * checking `meta = DandisetMeta(meta)` or `meta = PublishedDandisetMeta (meta)` (if `publishedBy` and `datePublished` is available):
      - if fields are missing (e.g. `contributors`) it would return error even today
      - I would add checks if `doi`  (and possibly other fields) is available
*  running `datacite_json = to_datacite(meta)`
      - this should not return an error

Author: On the previous design doc we were using the GitHub suggestion feature to inject changes rather than writing PRs for the PR:
```suggestion
* checking meta = DandisetMeta(meta) or meta = PublishedDandisetMeta (meta) (if publishedBy and datePublished is available):
  * if fields are missing (e.g. contributors) it would return error even today
  * I would add additional checks if doi (and possibly other fields) is available
* running datacite_json = to_datacite(meta)
  * this should not return an error
```

> TODO: what exceptions might be thrown? Missing creator is one.

Reviewer: missing creator should never happen at this stage. any metadata should be valid before the time this function is called.

Reviewer 3: ok

Author: ```suggestion
```

# Publish process

This process is predicated on a few tasks that are expected to receive valid input and assuming that only certain errors could happen that are out of our control.

- publish enabled only if dandiset is valid (so all validation checks have passed)
- on publish trigger:
   - create the versioned release in the db (this prevents any asset/blob belonging to the release from being deleted)
   - insert all the relevant metadata (version, doi, url, manifest location/checksum) into the metadata record of the new version 
  - send this new metadata record to mint the doi (there should be no reason why this should fail unless the external resource fails. if it fails notify admins since this is likely a new bug)
   - write out to s3 bucket (even on any datacite failure)

>    - switch over ownership of the release to dandi (unless releases are not modifiable by the dandiset creator)

Author: @satra Can you clarify this? "Owners" in `dandi-api` means a set of users who allowed to make changes through the API to a dandiset. Ownership only applies to dandisets as a whole, not versions. There's also "ownership" of the objects in the S3 bucket, but all objects belong to the sponsored bucket so I don't think that's what is meant here.

Do you mean that owners should not be allowed to modify published versions through the API?

Reviewer: correct. the only people who should be able to call PATCH/DELETE/PUT on non-draft dandisets are admins.

# Publish process

This process is predicated on a few tasks that are expected to receive valid input and assuming that only certain errors could happen that are out of our control.
Dandisets have a new attribute added to track their validation status: `VALID`, `INVALID`, and `VALIDATING`.
Whenever a dandiset is modified (metadata changed or files uploaded/removed), it is set to `VALIDATING` and an asynchronous task is triggered to run metadata validation on it. The task will update the state of the dandiset accordingly when it finishes.

- publish enabled only if a dandiset is `VALID`. The button will be greyed out in the UI and the API endpoint will return 405.
- on publish trigger:
   - create the versioned release in the db (this prevents any asset/blob belonging to the release from being deleted)

>    - insert all the relevant metadata (version, doi, url, manifest location/checksum) into the metadata record of the new version 

Reviewer: ```suggestion
   - insert all the relevant metadata (version, doi, url, manifest location) into the metadata record of the new version 
    -  manifest url == dandiarchive api url to GET on assets for the versioned dataset.
```

  - send this new metadata record to mint the doi (there should be no reason why this should fail unless the external resource fails. if it fails notify admins since this is likely a new bug)

>    - write out to s3 bucket (even on any datacite failure)

Reviewer: ```suggestion
   - write out to s3 bucket (even on any datacite failure)
      - add checksums for dandiset.yaml and assets.yaml which should be stored in a checksums.json in the published directory.
```

Author: I'm not clear on the value of checksumming the manifests. It doesn't provide any kind of security since it's stored in the same place as the content being checksummed. It doesn't really give any additional data integrity during the download process that HTTPS doesn't already give. Hypothetically it could safe guard against the API server uploading corrupted data, but this is not the best way to solve that problem.

Could you clarify what this is supposed to achieve?

Reviewer: i'm not sure aws provides checksums for things. this would mostly be a way for any user to verify that the dandiset.yaml or the assets.yaml have not changed. if the aws header already provides this, great, otherwise it's just another file containing the checksums of the files. i guess the AWS e-tag would be sufficient for the moment.

   
## DOI generation process

## dandi-api

The DOI generation process happens inside a call to `POST /dandisets/{dandiset_id}/draft/publish/`.
DOI generation is the first step of the publish process, because it is the only step to rely on an external service.
If the call to the Datacite API fails, there are no files in S3 that need to be cleaned up.

The API will pass the following to the DOI generator:

> * A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/dandi.000001.v0.210402.1835`.

Reviewer 3: assuming that we are using "." instead of "/" on purpose?  Should we change the form to something like this?
`f'{prefix}/dandi.{dandiset_id}.v.{version_id}'

* A URL for the dandiset. This will be a redirector URL of the form `f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'`.
* The dandiset metadata, as a python dictionary.

> ## dandi.core.doi ?

Reviewer 3: ```suggestion
## dandi.core.to_datacite
```

Reviewer 3: please see https://github.com/dandi/dandi-cli/pull/596

This function will generate the body of the DOI and return it.

> * checking meta = DandisetMeta(meta) or meta = PublishedDandisetMeta (meta) (if publishedBy and datePublished is available):
>   * if fields are missing (e.g. contributors) it would return error even today
>   * I would add additional checks if doi (and possibly other fields) is available
> * running datacite_json = to_datacite(meta)

Reviewer 3: ```suggestion
* creating a mapping between datacite's metadata and datacite's dictionary
```

>   * this should not return an error

Reviewer 3: ```suggestion
  * this should not return an error due to the problems with metadata (should be validated before)
```

## dandi-api

* If DOI generation raised any errors, return a 400 error describing the problem.
* Otherwise, attempt to register the DOI identifier/body with Datacite.

> * If that raises any errors, log them so they are reported in Sentry. Regardless of the outcome, continue with the publish process.

Author: ```suggestion
* If that raises any errors, log them so they are reported in Sentry. Regardless of the outcome, continue with the publish process.

## Validation process

### Assets
* Whenever an asset is created (I think this is the only way to set asset metadata), it kicks off an out of band job to validate the asset metadata.
* The task will grab the appropriate schema for the schema version (currently just the normal `asset.json`, but hopefully soon it will use a more restrictive `publishAsset.json`) and verify that the asset metadata validates against that schema. The result is stored on the asset.
* If it fails to validate, any draft dandisets which contain that asset will show in the API as `INVALID`, even if they would validate otherwise.

### Dandisets
* Whenever a new dandiset metadata is created, it kicks off an out of band job to validate the dandiset metadata. This is identical to the asset process, but uses the dandisets JSON schema.
* If any of the assets in a dandiset are `INVALID`, the dandiset is also `INVALID`. 

```

# Publish process

This process is predicated on a few tasks that are expected to receive valid input and assuming that only certain errors could happen that are out of our control.

> 
> - publish enabled only if dandiset is valid (so all validation checks have passed)

Author: ```suggestion
Dandisets have a new attribute added to track their validation status: `VALID`, `INVALID`, and `VALIDATING`.
Whenever a dandiset is modified (metadata changed or files uploaded/removed), it is set to `VALIDATING` and an asynchronous task is triggered to run metadata validation on it. The task will update the state of the dandiset accordingly when it finishes.

- publish enabled only if a dandiset is `VALID`. The button will be greyed out in the UI and the API endpoint will return 405.
```

If the call to the Datacite API fails, there are no files in S3 that need to be cleaned up.

The API will pass the following to the DOI generator:
* A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/dandi.000001.v0.210402.1835`.
* A URL for the dandiset. This will be a redirector URL of the form `f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'`.
* The dandiset metadata, as a python dictionary.

## dandi.core.doi ?

This function will generate the body of the DOI and return it.

* checking meta = DandisetMeta(meta) or meta = PublishedDandisetMeta (meta) (if publishedBy and datePublished is available):
  * if fields are missing (e.g. contributors) it would return error even today
  * I would add additional checks if doi (and possibly other fields) is available
* running datacite_json = to_datacite(meta)
  * this should not return an error

## dandi-api

* If DOI generation raised any errors, return a 400 error describing the problem.
* Otherwise, attempt to register the DOI identifier/body with Datacite.

> * If that raises any errors, log them and return a 400 error.
> * Finally, continue with the publish process.

Author: ```suggestion
* If that raises any errors, log them so they are reported in Sentry. Regardless of the outcome, continue with the publish process.
```

# Publish process

This process is predicated on a few tasks that are expected to receive valid input and assuming that only certain errors could happen that are out of our control.
Dandisets have a new attribute added to track their validation status: `VALID`, `INVALID`, and `VALIDATING`.
Whenever a dandiset is modified (metadata changed or files uploaded/removed), it is set to `VALIDATING` and an asynchronous task is triggered to run metadata validation on it. The task will update the state of the dandiset accordingly when it finishes.

- publish enabled only if a dandiset is `VALID`. The button will be greyed out in the UI and the API endpoint will return 405.
- on publish trigger:
   - create the versioned release in the db (this prevents any asset/blob belonging to the release from being deleted)
   - insert all the relevant metadata (version, doi, url, manifest location/checksum) into the metadata record of the new version 
  - send this new metadata record to mint the doi (there should be no reason why this should fail unless the external resource fails. if it fails notify admins since this is likely a new bug)
   - write out to s3 bucket (even on any datacite failure)

>    - switch over ownership of the release to dandi (unless releases are not modifiable by the dandiset creator)

Author: ```suggestion
```

Explicit switching of ownership isn't a thing, published versions are not modifiable.

The API will pass the following to the DOI generator:
* A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/dandi.000001.v0.210402.1835`.
* A URL for the dandiset. This will be a redirector URL of the form `f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'`.
* The dandiset metadata, as a python dictionary.

## dandi.core.doi ?

This function will generate the body of the DOI and return it.

* checking meta = DandisetMeta(meta) or meta = PublishedDandisetMeta (meta) (if publishedBy and datePublished is available):
  * if fields are missing (e.g. contributors) it would return error even today
  * I would add additional checks if doi (and possibly other fields) is available
* running datacite_json = to_datacite(meta)
  * this should not return an error

## dandi-api

* If DOI generation raised any errors, return a 400 error describing the problem.
* Otherwise, attempt to register the DOI identifier/body with Datacite.
* If that raises any errors, log them so they are reported in Sentry. Regardless of the outcome, continue with the publish process.

## Validation process

### Assets

> * Whenever an asset is created (I think this is the only way to set asset metadata), it kicks off an out of band job to validate the asset metadata.

Reviewer: ```suggestion
* Whenever an asset is created (I think this is the only way to set asset metadata), it kicks off an out of band job to validate the asset metadata. The asset metadata should be augmented with whatever is needed to satisfy the publishasset requirements. Thus this task may involve several out of band operations (e.g, calculate sha256, add other required metadata).
```
